File: api/src/factories/blockchain-connection.factory.ts
import { BadRequestException } from '@nestjs/common';

import { BlockchainPlatform } from '@app/models/enums';
import { IBlockchainConnectionService } from '@app/models/interfaces';
import { DummyConnectionService } from '@app/modules/dummy/services/dummy.service';
import { HyperledgerFabricConnectionService } from '@app/modules/hyperledger-fabric/services';

export class BlockchainConnectionFactory {
  private static instances: Map<string, IBlockchainConnectionService> =
    new Map();

  private static connections: Map<string, unknown> = new Map();

  static getService<T = IBlockchainConnectionService>(
    platform: BlockchainPlatform,
  ): T {
    if (!this.instances.has(platform)) {
      switch (platform) {
        case BlockchainPlatform.HYPERLEDGER_FABRIC:
          this.instances.set(
            platform,
            new HyperledgerFabricConnectionService(),
          );
          break;
        case BlockchainPlatform.DUMMY:
          this.instances.set(platform, new DummyConnectionService());
          break;
        default:
          throw new BadRequestException('UNSUPPORTED_BLOCKCHAIN_PLATFORM');
      }
    }
    return this.instances.get(platform)! as T;
  }

  static async getConnection<T, R>(
    blockchainId: string,
    blockchainPlatform: BlockchainPlatform,
    parameters: R,
  ) {
    const service = this.getService(blockchainPlatform);

    if (!this.connections.has(blockchainId)) {
      this.connections.set(blockchainId, await service.connect(parameters));
    }

    const conn = this.connections.get(blockchainId) as T;

    return conn;
  }
}

File: api/src/main/java/br/edu/unijui/gca/api/factories/BlockchainConnectionFactory.java
package br.edu.unijui.gca.api.factories;

import br.edu.unijui.gca.api.dtos.HyperledgerFabricConfigDto;
import br.edu.unijui.gca.api.enums.BlockchainPlatform;
import br.edu.unijui.gca.api.exceptions.UnsupportedBlockchainPlatformException;
import br.edu.unijui.gca.api.interfaces.IBlockchainConnection;
import br.edu.unijui.gca.api.services.HyperledgerFabricService;
import org.springframework.stereotype.Component;

import java.util.EnumMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class BlockchainConnectionFactory {
    private final Map<BlockchainPlatform, IBlockchainConnection<?, ?>> instances = new EnumMap<>(BlockchainPlatform.class);

    private final Map<String, Object> connections = new ConcurrentHashMap<>();

    public Class<?> getConfigType(BlockchainPlatform platform) {
        if (Objects.requireNonNull(platform) == BlockchainPlatform.HYPERLEDGER_FABRIC) {
            return HyperledgerFabricConfigDto.class;
        }

        throw new UnsupportedBlockchainPlatformException();
    }


    @SuppressWarnings("unchecked")
    public <R, S> IBlockchainConnection<R, S> getInstance(BlockchainPlatform platform) {
        if (!instances.containsKey(platform)) {
            if (Objects.requireNonNull(platform) == BlockchainPlatform.HYPERLEDGER_FABRIC) {
                instances.put(BlockchainPlatform.HYPERLEDGER_FABRIC, new HyperledgerFabricService());
            } else {
                throw new UnsupportedBlockchainPlatformException();
            }
        }

        return (IBlockchainConnection<R, S>) instances.get(platform);
    }

    public Object getConnection(String blockchainId, BlockchainPlatform blockchainPlatform, Object parameters) {
        var service = getInstance(blockchainPlatform);
        return connections.computeIfAbsent(blockchainId, id -> service.connect(parameters));
    }
}
